### DISCUSSION TITLE
Validating credentials of service account using MinIO Client or API

### DISCUSSION THREAD

**Discussion by jsabater** [minio/minio#19065] [Q&A]
Hello everyone! First time here and [cross-posting from Reddit](https://www.reddit.com/r/minio/comments/1as5key/validating_credentials_of_service_account_using/), where I also seeked help.

On a MinIO installation (latest version as of 2024-02-14), I have a user and this user has a number of service accounts, one per customer (separate web application per customer).

I use Ansible to deploy and update the application into separate LXC containers. This playbook includes handing the MinIO service account and bucket of each customer. When creating a new customer from scratch, the service account does not exist, but it does when updating.

I would like to know whether there is a way for me to use `mc admin user avcacct info` (or similar) to attempt an authentication and check whether my current credentials are valid, therefore I don't need to re-create or modify the service account.

I have been browsing [the documentation](https://min.io/docs/minio/linux/reference/minio-mc-admin/mc-admin-user.html) but I could not find it.

If `mc` does not support such feature, Maybe via the API? If so, how could I achieve that?  I could use the "parent" user that "owns" the service account to validate myself when calling a certain method.

Thanks in advance.

**Comment by jsabater** [comment]
Anyone, please? :pray: I just cannot believe there is no method to authenticate a user (just authenticate it, maybe return something given we are already there).

**Reply by harshavardhana** [reply]
If what you believe doesn't exist, please send a pull request. We are happy to accept sensible contributions.

**Reply by jsabater** [reply]
Thanks for your feedback, @harshavardhana. I have created a [feature request](https://github.com/minio/minio/issues/19263).

**Reply by jsabater** [reply]
Please feel free to close this discussion once you have validated that, indeed, this would be a new feature.

**Comment by bpedersen2** [comment]
Check the docs: https://min.io/docs/minio/linux/reference/minio-mc.html#test-the-connection

**Reply by jsabater** [reply]
Thanks for the suggestion, @bpedersen2, but it is not the service/server connection I am trying to test, but whether a service account credentials are still valid. I can get information about the service account with the `mc admin user svcacct info` command, but that won't tell me whether the stored password is valid (it could have been changed by some other actor).

**Reply by bpedersen2** [reply]
`mc admin info myminio` will only return success if the authentication succeeds, so testing the connection is more than "the minio host is reachable".

**Reply by jsabater** [reply]
I just re-read the documentation and noticed that you can use an `ACCESSKEY` and `SECRETKEY`. For some reason I had always used a username and password (i.e., the "parent" user of the service accounts) and didn't realise those applied to the service account as well.

So, thank you very much for the help.

**Reply by jsabater** [reply]
Unfortunately, I can't seem to make the following work:

```
# export MC_HOST_myalias="https://<ACCESSKEY>:<SECRETKEY>@minio1.domain.com:9000"
# mcli admin info myalias
mcli: <ERROR> Unable to get service status. Access Denied.
```

For reference, my list of aliases:

```
# mcli alias list
[..]
minio1
  URL       : https://minio1.domain.com:9000
  AccessKey : minio
  SecretKey : <password>
  API       : s3v4
  Path      : auto
[..]
```

What am I doing wrong?

References:

* https://min.io/docs/minio/linux/reference/minio-mc.html#test-the-connection
* https://min.io/docs/minio/linux/reference/minio-mc/minio-client-settings.html

**Reply by jsabater** [reply]
Further testing led me to notice that I can run a `mcli ls` command but not a `mcli admin info` command:

```
# mcli alias set myalias https://minio1.domain.com:9000 myaccesskey SECRETKEY
Added `myalias` successfully.
root@minio1:~# mcli ls myalias
[2022-11-17 15:15:08 UTC]     0B myaccesskey-media/
[2023-09-13 10:38:06 UTC]     0B myaccesskey-static/
root@minio1:~# mcli admin info myalias
mcli: <ERROR> Unable to get service status. Access Denied.
```

As I explained before, I had never attempted the `mcli admin info` command with a service account, only with a user. Am I missing some permission? This is the access policy of the service account `myaccesskey`:

```
{
 "Version": "2012-10-17",
 "Statement": [
  {
   "Effect": "Allow",
   "Action": [
    "s3:*"
   ],
   "Resource": [
    "arn:aws:s3:::myaccesskey-media/*",
    "arn:aws:s3:::myaccesskey-static/*"
   ]
  }
 ]
}
```

Thanks.

**Reply by jsabater** [reply]
I presume the missing permission has to be this one: [admin:ServerInfo](https://min.io/docs/minio/linux/administration/identity-access-management/policy-based-access-control.html#policy-action.admin-ServerInfo). Correct?

If so, I would have to modify all existing service accounts (access keys) and include such permission in their policy, correct? Then I would be able to use `mc admin info` to check whether a service account can authenticate with a set of credentials...

**Reply by jsabater** [reply]
I have tried adding the following bit to the policy of the service account I am setting the alias with:

```
  {
   "Effect": "Allow",
   "Action": [
    "admin:ServerInfo"
   ],
   "Resource": [
    "arn:aws:s3:::*"
   ]
  }
```

But I am still unable to athenticate my service account to check its credentials:

```
# mcli alias set myalias https://minio1.domain.com:9000 mysvcacct super-secret-password
Added `myalias` successfully.

# mcli alias list
myalias  
  URL       : https://minio1.domain.com:9000
  AccessKey : mysvcacct
  SecretKey : super-secret-password
  API       : s3v4
  Path      : auto

root@minio1:~# mcli admin info myalias
mcli: <ERROR> Unable to get service status. Access Denied.
```

Any hints, please? :pray:

**Reply by jsabater** [reply]
Okay, so I think I found the reason why it was not working: the service account I was setting the alias with belongs to a user.

I found out because I went to the MinIO Web Console and created a new service account from scratch (no parent user), and I added the following policy to it:

```
{
 "Version": "2012-10-17",
 "Statement": [
  {
   "Effect": "Allow",
   "Action": [
    "admin:ServerInfo"
   ]
  }
 ]
}
```

Then I followed the same procedure described in previous comments (i.e., setting the alias, then running the `mcli amdin info myalias` command) and it worked.

So I went to the parent user I was talking about and added the `diagnostics` policy to it, just to do a quick check, and it worked.

Thus my question would be, is it fine to have permissions such as that to the parent user? Am I compromising the system somehow when the service accounts used in the websites have access to `admin:ServerInfo` though its parent user?

**Comment by jaume-andronautic** [comment]
For future reference, this is the BASH script I am executing in an `ansible.builtin.shell` task. I have removed all the variables and added some placeholders (hopefully without mistakes), but you'll get the idea anyhow:

```bash
# Set temporary alias
mcli alias set tmp-alias https://$(hostname -f):9000 svc-account "password"
if [ "$?" == "0" ]
then
    # Credentials are valid: clean up and exit
    mcli alias remove tmp-alias
else
    # Credentials invalid: check if service account exists
    mcli admin user svcacct info $(hostname) "svc-account"
    if [ "$?" == "0" ]
    then
        # Service account exists: rotate password
        mcli admin accesskey edit $(hostname) "svc-account" --secret-key "password"
    else
        # Service account does not exist: create it
        mcli admin accesskey create $(hostname) parent_user \
        --access-key "svc-account" \
        --secret-key "password" \
        --policy "/tmp/svcacct-policy-.json" \
        --name "Some name" \
        --description "Some description"
    fi
fi
```
